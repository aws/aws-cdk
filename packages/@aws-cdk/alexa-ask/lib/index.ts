// Alexa::ASK CloudFormation Resources:
export * from './ask.generated';
export * from './pipeline-actions';
