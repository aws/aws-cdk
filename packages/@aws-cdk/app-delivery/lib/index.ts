export * from './pipeline-deploy-stack-action';
