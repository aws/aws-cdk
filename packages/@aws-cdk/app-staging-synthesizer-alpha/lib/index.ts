export * from './default-staging-stack';
export * from './app-staging-synthesizer';
export * from './bootstrap-roles';
export * from './staging-stack';
