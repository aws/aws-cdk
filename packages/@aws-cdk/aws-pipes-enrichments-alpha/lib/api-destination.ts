import { EnrichmentParametersConfig, IEnrichment, IPipe, InputTransformation } from '@aws-cdk/aws-pipes-alpha';
import { IApiDestination } from 'aws-cdk-lib/aws-events';
import { IRole, PolicyStatement } from 'aws-cdk-lib/aws-iam';
import { CfnPipe } from 'aws-cdk-lib/aws-pipes';

/**
 * Properties for a ApiDestinationEnrichment
 */
export interface ApiDestinationEnrichmentProps {
  /**
   * The input transformation for the enrichment
   * @see https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-pipes-input-transformation.html
   * @default - None
   */
  readonly inputTransformation?: InputTransformation;

  /**
   * The headers that need to be sent as part of request invoking the EventBridge ApiDestination.
   *
   * @default - none
   */
  readonly headerParameters?: Record<string, string>;

  /**
   * The path parameter values used to populate the EventBridge API destination path wildcards ("*").
   *
   * @default - none
   */
  readonly pathParameterValues?: string[];

  /**
   * The query string keys/values that need to be sent as part of request invoking the EventBridge API destination.
   *
   * @default - none
   */
  readonly queryStringParameters?: Record<string, string>;
}

/**
 * An API Destination enrichment for a pipe
 */
export class ApiDestinationEnrichment implements IEnrichment {
  public readonly enrichmentArn: string;

  private readonly inputTransformation?: InputTransformation;
  private readonly headerParameters?: Record<string, string>;
  private readonly pathParameterValues?: string[];
  private readonly queryStringParameters?: Record<string, string>;

  constructor(private readonly destination: IApiDestination, props?: ApiDestinationEnrichmentProps) {
    this.enrichmentArn = destination.apiDestinationArn;
    this.inputTransformation = props?.inputTransformation;
    this.headerParameters = props?.headerParameters;
    this.queryStringParameters = props?.queryStringParameters;
    this.pathParameterValues = props?.pathParameterValues;
  }

  bind(pipe: IPipe): EnrichmentParametersConfig {

    const httpParameters: CfnPipe.PipeEnrichmentHttpParametersProperty | undefined =
      this.headerParameters ??
        this.pathParameterValues ??
        this.queryStringParameters
        ? {
          headerParameters: this.headerParameters,
          pathParameterValues: this.pathParameterValues,
          queryStringParameters: this.queryStringParameters,
        }
        : undefined;

    return {
      enrichmentParameters: {
        inputTemplate: this.inputTransformation?.bind(pipe).inputTemplate,
        httpParameters,
      },
    };
  }

  grantInvoke(pipeRole: IRole): void {
    pipeRole.addToPrincipalPolicy(new PolicyStatement({
      resources: [this.destination.apiDestinationArn],
      actions: ['events:InvokeApiDestination'],
    }));
  }
}

