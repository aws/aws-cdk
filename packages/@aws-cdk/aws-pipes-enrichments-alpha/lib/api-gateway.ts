import { EnrichmentParametersConfig, IEnrichment, IPipe, InputTransformation } from '@aws-cdk/aws-pipes-alpha';
import { IRestApi } from 'aws-cdk-lib/aws-apigateway';
import { IRole, PolicyStatement } from 'aws-cdk-lib/aws-iam';
import { CfnPipe } from 'aws-cdk-lib/aws-pipes';

/**
 * Properties for a ApiGatewayEnrichment
 */
export interface ApiGatewayEnrichmentProps {
  /**
   * The input transformation for the enrichment
   * @see https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-pipes-input-transformation.html
   * @default - None
   */
  readonly inputTransformation?: InputTransformation;

  /**
   * The method for API Gateway resource.
   *
   * @default '*' - ANY
   */
  readonly method?: string;

  /**
   * The path for the API Gateway resource.
   *
   * @default '/'
   */
  readonly path?: string;

  /**
   * The deployment stage for the API Gateway resource.
   *
   * @default - the value of `deploymentStage.stageName` of target API Gateway resource.
   */
  readonly stage?: string;

  /**
   * The headers that need to be sent as part of request invoking the API Gateway REST API.
   *
   * @default - none
   */
  readonly headerParameters?: Record<string, string>;

  /**
   * The path parameter values used to populate the API Gateway REST API path wildcards ("*").
   *
   * @default - none
   */
  readonly pathParameterValues?: string[];

  /**
   * The query string keys/values that need to be sent as part of request invoking the EventBridge API destination.
   *
   * @default - none
   */
  readonly queryStringParameters?: Record<string, string>;
}

/**
 * An API Gateway enrichment for a pipe
 */
export class ApiGatewayEnrichment implements IEnrichment {
  public readonly enrichmentArn: string;

  private readonly inputTransformation?: InputTransformation;
  private readonly headerParameters?: Record<string, string>;
  private readonly pathParameterValues?: string[];
  private readonly queryStringParameters?: Record<string, string>;

  constructor(private readonly restApi: IRestApi, props?: ApiGatewayEnrichmentProps) {
    this.enrichmentArn = restApi.arnForExecuteApi(
      props?.method,
      props?.path || '/',
      props?.stage || this.restApi.deploymentStage.stageName,
    );
    this.inputTransformation = props?.inputTransformation;
    this.headerParameters = props?.headerParameters;
    this.queryStringParameters = props?.queryStringParameters;
    this.pathParameterValues = props?.pathParameterValues;
  }

  bind(pipe: IPipe): EnrichmentParametersConfig {

    const httpParameters: CfnPipe.PipeEnrichmentHttpParametersProperty | undefined =
      this.headerParameters ??
        this.pathParameterValues ??
        this.queryStringParameters
        ? {
          headerParameters: this.headerParameters,
          pathParameterValues: this.pathParameterValues,
          queryStringParameters: this.queryStringParameters,
        }
        : undefined;

    return {
      enrichmentParameters: {
        inputTemplate: this.inputTransformation?.bind(pipe).inputTemplate,
        httpParameters,
      },
    };
  }

  grantInvoke(pipeRole: IRole): void {
    pipeRole.addToPrincipalPolicy(new PolicyStatement({
      resources: [this.enrichmentArn],
      actions: ['execute-api:Invoke'],
    }));
  }
}

