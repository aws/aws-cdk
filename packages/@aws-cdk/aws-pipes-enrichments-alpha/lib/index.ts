export * from './api-destination';
export * from './lambda';
export * from './stepfunctions';
