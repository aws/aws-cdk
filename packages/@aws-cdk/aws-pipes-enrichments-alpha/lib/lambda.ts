import { EnrichmentParametersConfig, IEnrichment, IPipe, InputTransformation } from '@aws-cdk/aws-pipes-alpha';
import { IRole } from 'aws-cdk-lib/aws-iam';
import { IFunction } from 'aws-cdk-lib/aws-lambda';

/**
 * Properties for a LambdaEnrichment
 */
export interface LambdaEnrichmentProps {
  /**
   * The input transformation for the enrichment
   * @see https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-pipes-input-transformation.html
   * @default - None
   */
  readonly inputTransformation?: InputTransformation;
}

/**
 * A Lambda enrichment for a pipe
 */
export class LambdaEnrichment implements IEnrichment {
  public readonly enrichmentArn: string;

  private readonly inputTransformation?: InputTransformation;

  constructor(private readonly lambda: IFunction, props?: LambdaEnrichmentProps) {
    this.enrichmentArn = lambda.functionArn;
    this.inputTransformation = props?.inputTransformation;
  }

  bind(pipe: IPipe): EnrichmentParametersConfig {
    return {
      enrichmentParameters: {
        inputTemplate: this.inputTransformation?.bind(pipe).inputTemplate,
      },
    };
  }

  grantInvoke(pipeRole: IRole): void {
    this.lambda.grantInvoke(pipeRole);
  }
}

