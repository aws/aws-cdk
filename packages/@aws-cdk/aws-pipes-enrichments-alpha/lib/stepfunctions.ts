import { EnrichmentParametersConfig, IEnrichment, IPipe, InputTransformation } from '@aws-cdk/aws-pipes-alpha';
import { IRole } from 'aws-cdk-lib/aws-iam';
import { IStateMachine, StateMachine, StateMachineType } from 'aws-cdk-lib/aws-stepfunctions';

/**
 * Properties for a StepFunctionsEnrichment
 */
export interface StepFunctionsEnrichmentProps {
  /**
   * The input transformation for the enrichment
   * @see https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-pipes-input-transformation.html
   * @default - None
   */
  readonly inputTransformation?: InputTransformation;
}

/**
 * A StepFunctions enrichment for a pipe
 */
export class StepFunctionsEnrichment implements IEnrichment {
  public readonly enrichmentArn: string;

  private readonly inputTransformation?: InputTransformation;
  constructor(private readonly stateMachine: IStateMachine, props?: StepFunctionsEnrichmentProps) {
    if (stateMachine instanceof StateMachine
      && (stateMachine.stateMachineType !== StateMachineType.EXPRESS)
    ) {
      throw new Error(`EventBridge pipes only support EXPRESS workflows as enrichment, got ${stateMachine.stateMachineType}`);
    }
    this.enrichmentArn = stateMachine.stateMachineArn;
    this.inputTransformation = props?.inputTransformation;
  }

  bind(pipe: IPipe): EnrichmentParametersConfig {
    return {
      enrichmentParameters: {
        inputTemplate: this.inputTransformation?.bind(pipe).inputTemplate,
      },
    };
  }

  grantInvoke(pipeRole: IRole): void {
    this.stateMachine.grantStartSyncExecution(pipeRole);
  }
}

