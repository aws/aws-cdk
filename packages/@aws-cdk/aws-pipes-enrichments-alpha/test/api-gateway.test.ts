import { DynamicInput, InputTransformation, Pipe } from '@aws-cdk/aws-pipes-alpha';
import { App, Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { LambdaRestApi, RestApi } from 'aws-cdk-lib/aws-apigateway';
import { TestSource, TestTarget } from './test-classes';
import { ApiGatewayEnrichment } from '../lib/api-gateway';

describe('api-destination', () => {
  let app: App;
  let stack: Stack;
  let restApi: LambdaRestApi;

  beforeEach(() => {
    app = new App();
    stack = new Stack(app, 'TestStack');

    restApi = new RestApi(stack, 'RestApi');
    restApi.root.addResource('test').addMethod('GET');

  });

  it('should have only enrichment arn', () => {
    // ARRANGE
    const enrichment = new ApiGatewayEnrichment(restApi);

    new Pipe(stack, 'MyPipe', {
      source: new TestSource(),
      enrichment,
      target: new TestTarget(),
    });

    // ACT
    const template = Template.fromStack(stack);

    // ASSERT
    template.hasResourceProperties('AWS::Pipes::Pipe', {
      Enrichment: {
        'Fn::Join': [
          '',
          [
            'arn:',
            { Ref: 'AWS::Partition' },
            ':execute-api:',
            { Ref: 'AWS::Region' },
            ':',
            { Ref: 'AWS::AccountId' },
            ':',
            { Ref: 'RestApi0C43BF4B' },
            '/',
            { Ref: 'RestApiDeploymentStageprod3855DE66' },
            '/*/',
          ],
        ],
      },
      EnrichmentParameters: {},
    });
  });

  it('should have only enrichment arn with api gateway parameters', () => {
    // ARRANGE
    const enrichment = new ApiGatewayEnrichment(restApi, {
      stage: 'dev',
      method: 'GET',
      path: '/test',
    });

    new Pipe(stack, 'MyPipe', {
      source: new TestSource(),
      enrichment,
      target: new TestTarget(),
    });

    // ACT
    const template = Template.fromStack(stack);

    // ASSERT
    template.hasResourceProperties('AWS::Pipes::Pipe', {
      Enrichment: {
        'Fn::Join': [
          '',
          [
            'arn:',
            { Ref: 'AWS::Partition' },
            ':execute-api:',
            { Ref: 'AWS::Region' },
            ':',
            { Ref: 'AWS::AccountId' },
            ':',
            { Ref: 'RestApi0C43BF4B' },
            '/dev/GET/test',
          ],
        ],
      },
      EnrichmentParameters: {},
    });
  });

  it('should have enrichment parameters', () => {
    // ARRANGE
    const enrichment = new ApiGatewayEnrichment(restApi, {
      inputTransformation: InputTransformation.fromObject({
        body: DynamicInput.fromEventPath('$.body'),
      }),
      headerParameters: {
        headerParam: 'headerParam',
      },
      pathParameterValues: ['pathParam'],
      queryStringParameters: {
        param: 'queryParam',
      },
    });

    new Pipe(stack, 'MyPipe', {
      source: new TestSource(),
      enrichment,
      target: new TestTarget(),
    });

    // ACT
    const template = Template.fromStack(stack);

    // ASSERT
    template.hasResourceProperties('AWS::Pipes::Pipe', {
      EnrichmentParameters: {
        InputTemplate: '{"body":<$.body>}',
        HttpParameters: {
          HeaderParameters: {
            headerParam: 'headerParam',
          },
          PathParameterValues: [
            'pathParam',
          ],
          QueryStringParameters: {
            param: 'queryParam',
          },
        },
      },
    });
  });

  it('should grant pipe role invoke access', () => {
    // ARRANGE
    const enrichment = new ApiGatewayEnrichment(restApi);

    new Pipe(stack, 'MyPipe', {
      source: new TestSource(),
      enrichment,
      target: new TestTarget(),
    });

    // ACT
    const template = Template.fromStack(stack);

    // ASSERT
    expect(template.findResources('AWS::IAM::Role')).toMatchSnapshot();
    expect(template.findResources('AWS::IAM::Policy')).toMatchSnapshot();
  });
});

