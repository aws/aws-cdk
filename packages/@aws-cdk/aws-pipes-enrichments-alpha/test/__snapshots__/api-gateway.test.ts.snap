// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api-destination should grant pipe role invoke access 1`] = `
{
  "MyPipeRoleCBC8E9AB": {
    "Properties": {
      "AssumeRolePolicyDocument": {
        "Statement": [
          {
            "Action": "sts:AssumeRole",
            "Effect": "Allow",
            "Principal": {
              "Service": "pipes.amazonaws.com",
            },
          },
        ],
        "Version": "2012-10-17",
      },
    },
    "Type": "AWS::IAM::Role",
  },
  "RestApiCloudWatchRoleE3ED6605": {
    "DeletionPolicy": "Retain",
    "Properties": {
      "AssumeRolePolicyDocument": {
        "Statement": [
          {
            "Action": "sts:AssumeRole",
            "Effect": "Allow",
            "Principal": {
              "Service": "apigateway.amazonaws.com",
            },
          },
        ],
        "Version": "2012-10-17",
      },
      "ManagedPolicyArns": [
        {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
            ],
          ],
        },
      ],
    },
    "Type": "AWS::IAM::Role",
    "UpdateReplacePolicy": "Retain",
  },
}
`;

exports[`api-destination should grant pipe role invoke access 2`] = `
{
  "MyPipeRoleDefaultPolicy31387C20": {
    "Properties": {
      "PolicyDocument": {
        "Statement": [
          {
            "Action": "execute-api:Invoke",
            "Effect": "Allow",
            "Resource": {
              "Fn::Join": [
                "",
                [
                  "arn:",
                  {
                    "Ref": "AWS::Partition",
                  },
                  ":execute-api:",
                  {
                    "Ref": "AWS::Region",
                  },
                  ":",
                  {
                    "Ref": "AWS::AccountId",
                  },
                  ":",
                  {
                    "Ref": "RestApi0C43BF4B",
                  },
                  "/",
                  {
                    "Ref": "RestApiDeploymentStageprod3855DE66",
                  },
                  "/*/",
                ],
              ],
            },
          },
        ],
        "Version": "2012-10-17",
      },
      "PolicyName": "MyPipeRoleDefaultPolicy31387C20",
      "Roles": [
        {
          "Ref": "MyPipeRoleCBC8E9AB",
        },
      ],
    },
    "Type": "AWS::IAM::Policy",
  },
}
`;
